### ISSUE TITLE
Fine-tuning on cancer Xenium 5K data: training loss keeps increasing and entropy stays around 1.0

### ISSUE THREAD

**Issue by OMIC-coding** [MICS-Lab/novae#70]
Hi Novae team,

Thank you for developing Novae!

I am currently applying Novae to a cohort of cancer samples profiled with Xenium 5K. I am trying to fine-tune the pretrained human Novae model across multiple tumor and matched adjacent-tissue slides.

However, I noticed that the fine-tuning does not seem to converge properly. The `train/loss_epoch` does not decrease; instead, it gradually increases during training. At the same time, `train/entropy` stays very close to 1.0 throughout training.

For example, I observed something like:

- Epoch 4: `train/loss_epoch ~ 5.0`, `train/entropy ~ 1.0`
- Epoch 7: `train/loss_epoch ~ 6.2`, `train/entropy ~ 1.0`

The training then stops early because of early stopping.

My current setup is approximately:

```python
model.fine_tune(
    adatas,
    accelerator="gpu",
    min_prototypes_ratio=0.2,
    use_scheduler=True,
    num_workers=8,
)

**Comment by quentinblampey** [comment]
Hi @OMIC-coding!

Which Novae version do you use? The latest version (`novae==1.1.1`) should be more stable since we added a learning rate scheduler.

You can also try [this model](https://prism-oncology.github.io/novae/tutorials/novae_scConcept/), it should work better.

Other questions that may help: how many slides do you have? And how many cells in total?

**Comment by OMIC-coding** [comment]
Hi, thanks for your reply!

I am already using Novae v1.1.1.

My dataset consists of 100 tumor and 50 matched adjacent Xenium 5K samples. Each sample contains roughly 700,000 cells, so the full cohort is quite large and highly heterogeneous.

The main issue I am seeing during fine-tuning is that `train/loss_epoch` does not decrease and sometimes even gradually increases, while `train/entropy` remains very close to 1.0 throughout training. Reducing the learning rate did not seem to improve this behavior.

Do you have any idea what could cause this? Could it be related to the very large number of cells/slides, the high heterogeneity of tumor samples, or the prototype initialization?

Would you recommend fine-tuning on only a representative subset of slides first, or using `reference="all"` for prototype initialization?

Thanks again!

**Comment by quentinblampey** [comment]
> The main issue I am seeing during fine-tuning is that train/loss_epoch does not decrease and sometimes even gradually increases, while train/entropy remains very close to 1.0

Both show that the training is not going well, indeed. The entropy being close to 1 is even worse (everything above 0.9 is bad for sure).

The large number of cells/slides is not a problem at all, but it may influence the choice of the learning rate. You mentioned you reduced the learning rate, but how much? Given the size of your dataset, I suggest trying 1e-5 or even lower values.
For the `reference` parameter, keeping the default value (`None`) is usually a good decision (and also less arbitrary).

My best guess is that some slides are lower quality than others, but of course, I don't know the exact details and quality of your dataset... I suggest fine-tuning on a selected set of top-quality slides, e.g., 10 or 20 representative slides.

And, again, the [Novae+scConcept](https://prism-oncology.github.io/novae/tutorials/novae_scConcept/) model is worth trying: I noticed it's both more performant and more stable. On our proprietary datasets of 50M cells, we just use the default fine-tuning parameters (on the top quality slides) and it works very well.

**Comment by OMIC-coding** [comment]
Hi, thanks a lot for the detailed suggestions!!!

I’ll try a much lower learning rate (e.g. 1e-5 or below), and I’ll also test fine-tuning on a smaller set of high-quality slides. I will definitely try the Novae+scConcept model as well.

One thing I’m still wondering about is the tumor heterogeneity in my cohort. The cancer samples are highly heterogeneous across patients, and one of my main goals is to use the Novae-derived spatial representations/domains for downstream molecular stratification of the tumors.

In this setting, do you think fine-tuning only on 10–20 top-quality representative slides could bias the model toward the states present in those selected samples? In other words, would it be a concern that the model has not seen all tumor samples during fine-tuning, especially if some patients contain rare or patient-specific tumor states?

Would you recommend:
1. fine-tuning only on high-quality representative slides and then applying the model to the entire cohort, or
2. including as much of the tumor diversity as possible during fine-tuning, even if some slides are lower quality?

I’m particularly interested in preserving rare tumor-specific spatial programs while still obtaining a representation that is comparable across patients.

Thanks again!

**Comment by quentinblampey** [comment]
Sounds good!

I think you can start with (1) to at least have one run with a stable fine-tuning, and then you can try to add more slides to add more tumor diversity.

Note that you can use the `min_prototypes_ratio` parameter in the [`fine_tune` method](https://prism-oncology.github.io/novae/api/Novae/#novae.Novae.fine_tune) to decide how many niches should overlap across samples. By default, `min_prototypes_ratio=0.3`, meaning Novae will try to find at least 30% of shared niches per slide. You can play with this value depending on the heterogeneity of your data (although 0.3 should be a good start).

**Comment by OMIC-coding** [comment]
Hi, thanks again for your help!

I have one more question regarding zero-shot inference on a large number of slides.

When I run `compute_representations()` on multiple Xenium slides, the representation step itself works well for all slides. However, the kernel consistently crashes at the final step when Novae updates the prototypes with: `reference="all"`

My server has 2 TB of RAM, so I would expect the memory capacity to be sufficient even for a large cohort. The crash seems to happen specifically during the prototype computation / K-means step rather than during representation inference.

Since you mentioned that you have successfully used Novae on proprietary datasets with tens of millions of cells, I was wondering how you handle this step in practice.

Do you use `reference="all"` when computing zero-shot prototypes on such large datasets, or do you typically use only a subset of slides/cells as the reference?

Are there any settings that should be adjusted for very large datasets, for example:
- limiting BLAS/OpenMP threads,
- subsampling cells for prototype initialization,
- changing the K-means settings,
- using a subset of representative slides,
- or any other parameter related to prototype computation?

In my case, the representations are computed successfully, but the kernel crashes specifically when the prototypes are updated from all reference cells.

I would really appreciate any advice on how you normally perform prototype computation at this scale.

Thanks again!

**Comment by quentinblampey** [comment]
Hi,

Indeed, in zero-shot, the K-Means can cause some issues with a large number of slides. Usually, this can be mitigated with one of the equivalent solutions below:
1. You first run zero-shot on some selected slides of interest, and then run compute_representations on the other slides (with `zero_shot=False` to not update the prototypes)
2. Or, you run zero-shot on all slides, but select the slides of interest using `reference` (list of slide indices or slide IDs)
Both are essentially doing the same thing, you can pick the one that is more convenient.

This is just a workaround though...
I think I can sample 1M representations before running the K-Means to avoid the crash. Sampling 1M representations shouldn't have a big impact on the K-Means anyway. What do you think? I opened an issue [here](https://github.com/prism-oncology/novae/issues/71)
